// @prismicio/api-renderer/src/widgets/nestable/Link/MediaLinkRenderer.ts
import type { LinkContent, MediaLinkContent } from "@prismicio/types-internal/lib/content"

import type { LinkRenderer, LinkRendererOptions, MediaLinkDef } from "../../../models"

const MediaLinkRenderer: LinkRenderer<
	MediaLinkDef,
	LinkContent & { value: MediaLinkContent },
	undefined,
	LinkRendererOptions
> = {
	renderV1(link: LinkContent & { value: MediaLinkContent }, _fetch: undefined, options?: LinkRendererOptions): unknown {
		return {
			type: "Link.media",
			value: {
				...(!options?.omitKey && { key: link.key }),
				...(link.value.text && { text: link.value.text }),
			},
		}
	},

	renderV2(
		def: MediaLinkDef,
		link: LinkContent & { value: MediaLinkContent },
		_fetch: undefined,
		options?: LinkRendererOptions,
	): unknown {
		return {
			link_type: "Media",
			...(!options?.omitKey && { key: link.key }),
			...(def.allowText && link.value.text && { text: link.value.text }),
		}
	},

	renderMocks(
		def: MediaLinkDef,
		link: LinkContent & { value: MediaLinkContent },
		options?: LinkRendererOptions,
	): unknown {
		return this.renderV2(def, link, undefined, options)
	},
}

export default MediaLinkRenderer
